pkg/notify/notifier/pushover/pushover.go

```diff
@@ -0,0 +1,256 @@
+			return err
+		}
+
+		// send messages in parallel
+		g := new(errgroup.Group)
+		for _, message := range messages {
+			msg := message
+			g.Go(func() (err error) {
+				// consume a token from limiter bucket
+				if err := n.limiter.Wait(context.Background()); err != nil {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: limiter error", "userKey", userKey, "error", err.Error())
+					return err
+				}
+
+				// construct pushover message struct as request parameters, and validate it
+				pm := newPushoverMessageExtend(token, userKey, msg, *profile.Title, *profile.Sound, profile.Devices)
+				err, warnings := pm.validate()
+				if err != nil {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: invalid pushover message", "userKey", userKey, "error", err.Error())
+					return err
+				}
+				if len(warnings) > 0 {
+					_ = level.Warn(n.logger).Log("msg", "PushoverNotifier: warnings about the message", "userKey", userKey, "warnings", strings.Join(warnings, "; "))
+				}
+				pReq := &pushoverRequest{pm}
+
+				// JSON encoding
+				var buf bytes.Buffer
+				if err := utils.JsonEncode(&buf, pReq); err != nil {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: encode message error", "userKey", userKey, "error", err.Error())
+					return err
+				}
+
+				// build a JSON request with context
+				request, err := http.NewRequestWithContext(context.Background(), http.MethodPost, URL, &buf)
+				if err != nil {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: encode http request error", "userKey", userKey, "error", err.Error())
+					return err
+				}
+				request.Header.Set("Content-Type", "application/json")
+
+				// send the request
+				response, err := n.client.Do(request)
+				if err != nil {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: do http error", "userKey", userKey, "error", err.Error())
+					return err
+				}
+
+				defer func() {
+					_, _ = io.Copy(ioutil.Discard, response.Body)
+					_ = response.Body.Close()
+				}()
+
+				// check status code, but not return error if it is not 2xx, since we will do this later
+				if response.StatusCode != http.StatusOK {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: got non-2xx response", "userKey", userKey, "StatusCode", response.StatusCode)
+					body, _ := ioutil.ReadAll(response.Body)
+					return fmt.Errorf("PushoverNotifier: got non-2xx response, StatusCode: %d, response: %s", response.StatusCode, string(body))
+				}
+
+				// check if the remaining number of messages that can be sent is low
+				PoRemainingMsg, err := strconv.Atoi(response.Header.Get("X-Limit-App-Remaining"))
+				if err != nil {
+					_ = level.Error(n.logger).Log("msg", "PushoverNotifier: get response headers error", "userKey", userKey, "error", err.Error())
+					return err
+				}
+				if PoRemainingMsg < PoMsgLimitAlert {
```

Comment:
It needs to continue sending messages or not when the message reached the limit?

Reply:
Here it is allowed to continue to send messages no matter how many quotas left. An error will be obtained from the response and returned if the Pushover app runs out of its message capacity.